Chargebee/Classes/Purchase/CBPurchaseManager.swift

```diff
@@ -262,43 +262,54 @@ extension CBPurchase: SKPaymentTransactionObserver {
-                    switch receiptResult {
-                    case .success(let receipt):
-                        debugPrint("Receipt: \(receipt)")
-                        if receipt.subscriptionId.isEmpty {
-                            completion?(.failure(CBError.defaultSytemError(statusCode: 400, message: "Invalid Purchase")))
-                            return
-                        }
-                        self.activeProduct = nil
-                        completion?(.success((true, receipt.subscriptionId, receipt.planId)))
-                    case .error(let error):
-                        debugPrint(" Chargebee - Receipt Upload - Failure")
-                        completion?(.failure(error))
+        
+        CBReceiptValidationManager.validateReceipt(receipt: receipt) {
+            (receiptResult) in DispatchQueue.main.async {
+                switch receiptResult {
+                case .success(let receipt):
+                    debugPrint("Receipt: \(receipt)")
+                    if receipt.subscriptionId.isEmpty {
+                        completion?(.failure(CBError.defaultSytemError(statusCode: 400, message: "Invalid Purchase")))
+                        return
                     }
+                    self.activeProduct = nil
+                    completion?(.success((true, receipt.subscriptionId, receipt.planId)))
+                case .error(let error):
+                    debugPrint(" Chargebee - Receipt Upload - Failure")
+                    completion?(.failure(error))
                 }
             }
-        } catch { print("Couldn't read receipt data with error: " + error.localizedDescription) }
-
+        }
     }
+    
+    internal func getReceipt(product: SKProduct?) -> CBReceipt? {
+        var receipt: CBReceipt?
+         guard let appStoreReceiptURL = Bundle.main.appStoreReceiptURL,
+               FileManager.default.fileExists(atPath: appStoreReceiptURL.path) else {
+             debugPrint("No receipt Exist")
+             return nil
+         }
+         guard let product = product, let currencyCode = product.priceLocale.currencyCode, let period = product.subscriptionPeriod?.numberOfUnits, let unit = product.subscriptionPeriod?.unit.rawValue  else {
+             return nil
+         }
+         do {
+             let receiptData = try Data(contentsOf: appStoreReceiptURL, options: .alwaysMapped)
+             
+             let receiptString = receiptData.base64EncodedString(options: [])
+             debugPrint("Apple Purchase - success")
+              receipt = CBReceipt(name: product.localizedTitle, token: receiptString, productID: product.productIdentifier, price: "\(product.price)", currencyCode: currencyCode, period: period, periodUnit: Int(unit),customer: customer)
+         }catch {
+             print("Couldn't read receipt data with error: " + error.localizedDescription)
+         }
+        return receipt
+        
+     }
```

Comment:
Should we throw an error if there are no receipts found so that the user can decide on how to proceed?

Reply:
Also needs some minor formatting.

Reply:
Throwing error when receipt not valid or not found  while getting receipt 

   ```
 guard let receipt = getReceipt(product: product?.product) else {
            debugPrint("Couldn't read receipt data with error")
            completion?(.failure(CBError.defaultSytemError(statusCode: 0, message: "Could not read receipt data")))
            return
        }
```